lib/Db/Helper.php

```diff
@@ -38,13 +38,28 @@ public static function upsert(Connection $connection, string $table, array $data
 
             return $connection->insert($table, $data);
         } catch (\Doctrine\DBAL\Exception\UniqueConstraintViolationException $exception) {
-            $critera = [];
+            $criteria = [];
+            $checkKeys = [];
+            
             foreach ($keys as $key) {
-                $key = $quoteIdentifiers ? $connection->quoteIdentifier($key) : $key;
-                $critera[$key] = $data[$key] ?? throw new \LogicException(sprintf('Key "%s" passed for upsert not found in data', $key));
+                $quotedKey = $quoteIdentifiers ? $connection->quoteIdentifier($key) : $key;
+                $criteria[$quotedKey] = $data[$quotedKey] ?? $checkKeys[] = $key;
             }
 
-            return $connection->update($table, $data, $critera);
+            if ($checkKeys) {
+                $schemaManager = $connection->createSchemaManager();
+                $indexes = $schemaManager->listTableIndexes($table);
+                foreach ($checkKeys as $key) {
+                    foreach ($indexes as $index) {
+                        if ([$key] === $index->getColumns() && $index->isUnique()) {
```

Comment:
@kingjia90 What happens if there is a unique index with multiple columns and the key is only one of the columns?

Reply:
Afaik, all the columns must be considered by the composite unique index to trigger the unique contraint.

Reply:
But when I have for example 2 Objects with the same `path` and then I change the `key` that both have the same `key` and `path` and `key` are unique. Then we have a problem right?

Reply:
![image](https://github.com/pimcore/pimcore/assets/6014195/4ec903ff-97a0-4b61-92ca-13abf1cc0aa0)
mhhh, for sure, `fullpath` index would be triggered. This draft doesn't cover these cases yet at all, but i see it's gonna getting more and more complex nested loops 😟